README.md
# Titanic-Machine-Learning-from-Disaster
**Статус проекта:** планируются доработки

<img alt="Титаник" src="/sources/Titanic.jpg" height="500">


## Краткое описание
Решение для [Titanic ML competition](https://www.kaggle.com/c/titanic): анализ данных, модель, предсказания, результат на Kaggle.

**Цель** - сделать модель, которая, получив данные о пассажире Титаника, предскажет выжил
пассажир в кораблекрушении или нет.

Решение в [тетрадке](/sources/Titanic-Machine-Learning-from-Disaster.ipynb).


## Приёмы, которые демонстрирует проект
1. Исследовательский анализ данных
2. Тестирование статистических гипотез (гипотеза о равенстве средних - с помощью **scipy.stats.ttest_ind**)
3. Заполнение пропусков
4. Разработка новых признаков
5. Манипуляции с данными средствами **pandas**
6. Использование конвейеров (*Pipeline*) из **sklearn**
7. Оценка информативности признаков
7. Работа с моделями градиентных бустингов: **XGBoost**
8. Оценка моделей
9. Автоматизация оценки производительности моделей и вывода результатов (код собран в функции, устранены повторения кода)
10. Настройка их гиперпараметров с **Optuna**
11. Распараллеливание вычислений с помощью **joblib**
11. Запуск проекта совместно с базой данных с помощью *Docker Compose*
12. Оформление графиков с **seaborn** и **matplotlib**
    - График из исследовательского анализа:
    <img alt="График из исследовательского анализа" src="/img/plots_0.png" height="500">
    
    - Графики из раздела с про разработку признаков:
    <img alt="Матрица корреляции" src="/img/plots_1.png" height="900">
    <img alt="Взаимная информация" src="/img/plots_2.png" height="500">
    
    - Кривые ROC выбранных моделей после настройки гиперпараметров:
    <img alt="ROC curve" src="/img/plots_3.png" height="500">

13. Логирование с помощью чат-бота в Телеграм.
<img alt="логирование" src="/img/log.jpg" height="500">


## Результат на Kaggle
К сожалению, результат не дотянул до медианного.  
<img alt="Результат на Kaggle" src="/sources/scores_on_kaggle.PNG" height="70">

Во время работы над проектром было перепробовано несколько разных моделей, которые показывали на Kaggle целевую метрику в районе 0.77 - 0.78. Модель выбранная для оформления на GitHub неожиданно показала результат ниже. В дальнейшем она будет улучшена.

Возможные способы улучшения результата:
- удалить выбросы в данных;
- разработать более информативные признаки;
- избавиться от дисбаланса классов с помощью апсемплинга или даунсемплинга, учесть дисбаланс с помощью весов классов;
- продолжить настройку гиперпараметров и побороть оставшийся оверфитинг;
- попробовать использовать более мощный классификатор;
- попробовать объединить классификаторы в ансамбль.


## Запуск проекта
В проекте реализована оптимизация гиперпараметров с **Optuna**, распаралленная
на несколько потоков ("воркеров"), при этом, чтобы сократить количество 
устанавливаемых библиотект, **optuna-distributed** не использовалась. 
Это решение накладывает ограничения на запуск: необходимо сохранять 
промежуточные результаты оптимизации в базе данных.

К **Optuna** для сохранения промежуточных результатов подключается база данных
*PostgreSQL* в строке
```Python
optuna_storage = optuna.storages.RDBStorage(
    url='postgresql+psycopg2://optuna:password@localhost/optuna'
)
```
Пользователь должен самостоятельно установить и настроить базу данных и, если необходимо,
изменть настройки подключения при инициализации `RDBStorage`. Также можно подключить к 
**Optuna** любое другое хранилище, способное работать с многопоточной оптимизацией 
(*SQLite* - не очень хорошее решение), либо убрать распараллеливание из проекта и, при 
необходимости, хранилище с промежуточными результатами оптимизации.

Для запуска проекта в установленном дистрибутиве *Anaconda* необходимо:
- создать окружение из файла **requirements.yml** с помощью команды `conda env create -f requirements.yml` 
(предполагается, что коммандная строка запущена из каталога с **requirements.yml**), созданное окружение
будет называться `titanic`;
- активировать окружение с помощью команды `conda activate titanic`.  
После этого с проектом можно будет работать, например с помощью *jupyter notebook*;
- в командной строек появится сылка на сервер *jupyter*, необходимо прейти по ней, 
чтобы попасть на сервер *jupyter*.


## Логирование 
В проекте сделано автоматическое логирование в телеграм с **notifiers**. Параметры `notifier.notify`, в 
частности, токен и ID чата должны быть размещены в папке с тетрадкой в файле `notifier_params.pkl`.
Естественно, этот файл отсутствует в репозитории, поэтому для настройки логирования и запуска тетрадки
в исходном виде необходимо создать файл `notifier_params.pkl`, содержащий словарь с параметрами для
**notifiers**. Ниже пример словаря:
```Python
notifier_params = {
    'notifier': 'telegram',
    'token': '**********************************************',
    'chat_id': 0000000000
}
```

Также можно просто удалить все следы логирования и библиотеки **notifiers**.

## Структура репозитория
- **datasets** - набор данных и таблица с результатами, взятые на Kaggle
- **img** - иллюстрации для README.md
- **sources**
    - **Titanic-Machine-Learning-from-Disaster.ipynb** - Тетрадка с решением задачи
    - **Titanic.jpg** и **scores_on_kaggle.PNG** - иллюстрации для тетрадки
    - **submission.csv** - предсказания, отправленные на Kaggle
- **requirements.yml** - окружение для запуска проекта
- **LICENSE** - MIT License
- **README.md** - файл с описанием проекта


change_log.txt
2023-11-06 08:27
Созданы черновики с улучшенными моделями и улучшенными новыми признаками.

2023-11-06 14:29
1. Установлен flake8-nb
2. Файлы в папке sources оформлены так, чтобы линтер их пропустил
3. Тетрадки перезапущены, обновлена запись в mlflow.

2023-09-02 23:08
1. Локально установлены линтеры flake8 и mypy.
2. Добавлен tox.ini с минимальными настройками для flake8.
3. В requirements.yml экспортировано окружение с установленными линтерами.


2023-09-02 22:39
1. Добален pyproject.toml с настройками форматера black.
2. Файл sources/useful_funcs.py отформатирован в соответствии с новыми настройками.

2023-08-28 22:58
1. Установлен автоформатер Black для работы с jupyter
2. В requirements.yml экспортировано окружение с установленным автоформатером
3. Автоформатером приведено в порядок оформление файла sources/useful_funcs.py


2023-08-28 07:53
1. Удалены лишние импорты.

2023-08-18 23:15

1. Добавлено журналирование в MLflow.


2023-08-10 08:05

1. Исправлены незнечительные ошибки в текстах.


2023-08-09 23:10

1. Базовая модель вынесена в отдельный файл sources/Baseline.ipynb.


2023-08-09 07:22

1. Тренировочный набор данных сохранён в отдельный файл train_set.joblib, чтобы обеспечить
его постоянство в разных тетрадках.
2. Файлы с расширением .joblib добавлены в .gitignore.



2023-08-09 06:11

1. Удалены лишние импорты в тетрадках.



2023-08-08 21:26

1. В useful_funcs.py перенесены функции для оценки моделей:
- get_cv_scores,
- get_train_score,
- get_cv_pred,
- plot_roc_curve_for_random_clf,
- plot_roc_curve,
- agg_scores,
- score_box_plot,
- evaluate_model,
- compare_models.



2023-08-08 07:35

1. Исследовательский анализ перенесён в файл EDA.ipynb
2. Часть повторяющихся функций (make_notifier, plot_ecdf_with_target, check_target_imbalance, plot_corr_matrix) вынесены в отдельный модуль useful_funcs.py.
3. Создан файл change_log.txt, в который будет заноситься информация об изменениях в проекте.



sources/infographics.py
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import seaborn as sns

from sklearn.feature_selection import mutual_info_classif
from sklearn.metrics import (#accuracy_score,
                             #f1_score,
                             get_scorer, 
                             roc_curve,
                            )
from sklearn.model_selection import (cross_val_predict, 
                                     cross_val_score,
                                    )


def plot_corr_matrix(df, size=(7, 7), vmin=-1, vmax=1, method='pearson', annot=True):
    corr = df.corr(method=method)
    
    mask = np.triu(np.ones_like(corr, dtype=bool))
    
    f, ax = plt.subplots(figsize=size)
    
    cmap = sns.diverging_palette(230, 20, as_cmap=True)
    
    sns.heatmap(corr,
                mask=mask,
                cmap=cmap,
                annot=annot,
                vmax=vmax,
                vmin=vmin,
                center=0.0,
                square=True,
                linewidths=1.0,
                cbar_kws={'shrink': 0.5},
                ax=ax
               )
    
    ax.set_title('Матрица корреляции')
    
    plt.show()
    
    
def print_mi_scores(df, y_train, n_neighbors, random_state=None):
    
    fig, ax = plt.subplots()
    
    mi_scores = (pd
                 .DataFrame(mutual_info_classif(df, 
                                                y_train, 
                                                n_neighbors=n_neighbors, 
                                                random_state=random_state), 
                            columns=['mutual_info'], 
                            index=df.columns
                           )
                 .sort_values(by='mutual_info', ascending=False)
    )

#     print(f'n_neighbors={n_neighbors}')
#     display(mi_scores)

    (mi_scores
     .round(3)
     .sort_values(by='mutual_info', ascending=True)
     .plot(kind='barh', grid=False, title=f'Mutual information\nn_neighbors={n_neighbors}', ax=ax)
    )

    ax.bar_label(ax.containers[0])
    ax.set_xlim([0, 0.25])
    ax.legend(loc='lower right')
    plt.show()
    
    return mi_scores


def print_score(estimator, features, target, score='accuracy', cv=5, n_jobs=-1):
    '''Вычисляет среднее, минимальное и максимальное значения метрики,
    полученное с помощью кросс-валидации.
    '''
    
    scores = cross_val_score(estimator,
                             X=features,
                             y=target,
                             cv=cv,
                             scoring=score,
                             n_jobs=n_jobs,
                            )
    
    res = pd.DataFrame(scores, columns=[score]).agg(['mean', 'median', 'min', 'max']).transpose()
    
    scorer = get_scorer(score)
    res.loc[score, 'results_on_train_set'] = scorer(estimator.fit(features, target), features, target)
    
    return res


def plot_roc_curve_for_random_clf():
    '''Выводит на экран кривую ROC для классификатора, предсказывающего
       целевую переменную случайным образом.
    '''
    
    fig, ax = plt.subplots(1, 1)
    
    ax.plot([0,1], [0,1], 'k--', label='Random classifier')
    ax.grid()
    ax.set_xlim((0, 1))
    ax.set_ylim((0, 1))

    ax.set_xlabel('False Positive Rate')
    ax.set_ylabel('True Positive Rate')

    ax.legend(loc='lower right')

    ax.set_title('ROC curve')
    
    return fig, ax


def plot_roc_curve(y_train, y_scores, ax, style='', label=None):
    '''Печатае кривую ROC на оси ax.'''
    
    fpr, tpr, thresholds = roc_curve(y_train, y_scores)
    
    ax.plot(fpr, tpr, style, label=label)
    
    return ax


def evaluate_model(model, X, y, label, ax=None, method='predict_proba', cv=5, n_jobs=-1):
    '''Выводит на экран сводку о производительности модели, состоящую из метрик
    accuracy, f1, roc_auc и кривую ROC. Также возвращат значения метрик в датафрейме df,
    печатает кривую ROC на оси ax.
    '''
    
    df = pd.DataFrame(columns= ['mean', 'median', 'min', 'max'])
    
    for metric in ['accuracy', 'f1', 'roc_auc']:
        df = pd.concat([df, print_score(model, X, y, score=metric, cv=cv, n_jobs=n_jobs)])
        
    print(f'Значения метрик для {label}')
    
    display(df)
        
    y_scores = cross_val_predict(model, X, y, cv=cv, method=method, n_jobs=n_jobs)
    if method == 'predict_proba':
        y_scores = y_scores[:, -1]
    
    if ax is None:
        fig, ax = plot_roc_curve_for_random_clf()        
        ax = plot_roc_curve(y, y_scores, label=label, ax=ax)
        plt.show()
        
        return df
    
    else:
        ax = plot_roc_curve(y, y_scores, label=label, ax=ax)
        
        return df, ax
    
    
def plot_accuracy(df, aspect=1, rot=0):
    g = sns.catplot(aspect=aspect)

    g = sns.scatterplot(data=df, 
                      x='classifier', 
                      y='accuracy_on_train_set', 
                      markers='X', 
                      label='Accuracy на тренировочном наборе'
                     )

    g = sns.boxplot(data=df, x='classifier', y='accuracy_on_CV', color='white', showmeans=True)

    g.set_ylim([0.6, 1])
    g.tick_params(axis='x', rotation=rot)
    g.set_xlabel('Классификатор')
    g.set_ylabel('Accuracy')
    g.set_title('Значения метрики accuracy по результатм\nкросс-валидации для разных классификаторов')
    sns.move_legend(g, 'lower right')

    plt.show()
    
    
def get_performance_of_one_model(model, X, y, label, scoring='accuracy', cv=5, n_jobs=-1, method='predict_proba'):
    
    fig, ax = plot_roc_curve_for_random_clf()
    
    scores = pd.DataFrame(columns=['classifier', 'accuracy_on_CV', 'accuracy_on_train_set'])
    
    print(label)
    
    df, ax = evaluate_model(model=model, X=X, y=y, method=method, label=label, cv=cv, ax=ax)

    score = pd.DataFrame(cross_val_score(estimator=model, 
                                         X=X, 
                                         y=y, 
                                         scoring=scoring,
                                         cv=cv,
                                         n_jobs=n_jobs,
                                        ),
                         columns=['accuracy_on_CV']
                        )
    score['classifier'] = label
    score['accuracy_on_train_set'] = df.loc['accuracy', 'results_on_train_set']
    scores = scores.append(score)
    
    print() 
    
    plt.legend()
    plt.show()

    print()

    plot_accuracy(scores)

sources/useful_funcs.py
import datetime
import joblib
import notifiers
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from scipy.stats import percentileofscore
import seaborn as sns

from sklearn.metrics import (
    get_scorer,
    roc_curve,
)

from sklearn.model_selection import (
    cross_val_predict,
    cross_val_score,
)


def make_notifier():
    """Настраивает логирование в телеграм. Возвращает функцию
    для отправки сообщений.
    """

    logger_params = joblib.load("notifier_params.pkl")

    def notifier_func(text):
        now = datetime.datetime.now()
        text_with_dt = f'{now.strftime("%d-%m-%Y %H:%M:%S")}\n{text}'

        notifier = notifiers.get_notifier(logger_params["notifier"])
        notifier.notify(
            message=text_with_dt,
            token=logger_params["token"],
            chat_id=logger_params["chat_id"],
        )

    return notifier_func


def plot_ecdf_with_target(data, target):
    sns.displot(data, stat="proportion", kind="ecdf", height=5, aspect=1)

    quantile = percentileofscore(data, target) / 100

    plt.plot([0, target, target], [quantile, quantile, 0], "-.r")
    plt.plot([target], [quantile], "or")

    plt.xlim((0, 1))
    plt.ylim((0, 1))

    plt.title("ECDF результатов соревнования")

    plt.grid()

    plt.show()


def check_target_imbalance(vals):
    mean_val = vals.mean()

    print(f"Доля выживших пассажиров - {mean_val: .2%}")
    print(f"Доля погибших пассажиров - {1 - mean_val: .2%}")


def plot_corr_matrix(
    df, size=(7, 7), vmin=-1, vmax=1, method="pearson", annot=True
):
    corr = df.corr(method=method)

    mask = np.triu(np.ones_like(corr, dtype=bool))

    f, ax = plt.subplots(figsize=size)

    cmap = sns.diverging_palette(230, 20, as_cmap=True)

    sns.heatmap(
        corr,
        mask=mask,
        cmap=cmap,
        annot=annot,
        vmax=vmax,
        vmin=vmin,
        center=0.0,
        square=True,
        linewidths=1.0,
        cbar_kws={"shrink": 0.5},
        ax=ax,
    )

    ax.set_title("Матрица корреляции")

    plt.show()


def get_cv_scores(estimator, X, y, cv, n_jobs, score="accuracy"):
    """Выполняет кросс-валидацию и возвращает значения целевой метрики.
    По сути, обёртка для cross_val_score с заданными значениями параметров
    cv и n_jobs.
    """

    scores = cross_val_score(
        estimator,
        X=X,
        y=y,
        cv=cv,
        scoring=score,
        n_jobs=n_jobs,
    )

    return scores


def get_train_score(estimator, X, y, score="accuracy"):
    """Обучает модель на тренировочной выборке и на ней же оценивает
    заданную метрику.
    """

    scorer = get_scorer(score)

    return scorer(estimator.fit(X, y), X, y)


def get_cv_pred(estimator, X, y, cv, n_jobs, method="predict"):
    """Обёртка вокруг cross_val_predict, возвращающая массив с предсказаниями
    на каждом из фолдов.
    """

    y_pred = cross_val_predict(
        estimator, X, y, cv=cv, method=method, n_jobs=n_jobs
    )

    if method == "predict_proba":
        y_pred = y_pred[:, -1]

    return y_pred


def plot_roc_curve_for_random_clf():
    """Выводит на экран кривую ROC для классификатора, предсказывающего
    целевую переменную случайным образом.
    """

    fig, ax = plt.subplots(1, 1)

    ax.plot([0, 1], [0, 1], "k--", label="Random classifier")
    ax.grid()
    ax.set_xlim((0, 1))
    ax.set_ylim((0, 1))

    ax.set_xlabel("False Positive Rate")
    ax.set_ylabel("True Positive Rate")

    ax.legend(loc="lower right")

    ax.set_title("ROC curve")

    return fig, ax


def plot_roc_curve(y_train, y_scores, label, ax, style=""):
    """Печатае кривую ROC на оси ax."""

    fpr, tpr, thresholds = roc_curve(y_train, y_scores)

    ax.plot(fpr, tpr, style, label=label)

    ax.legend(loc="lower right")

    return ax


def agg_scores(scores, label, score_names):
    """Возвращает датафрейм, содержащий среднее, медианное, минимальное и
    максимальное значения метрики, полученное с помощью кросс-валидации, и
    оценку метрики на тренировочном наборе.
    """

    res = (
        scores.query("classifier == @label")[score_names]
        .agg(["min", "median", "mean", "max"])
        .transpose()
        .rename(columns=lambda string: "_".join(["cv", string]))
    )

    cols_with_train_scores = [
        "train_" + score_name for score_name in score_names
    ]
    train_score = scores.query("classifier == @label").loc[
        0, cols_with_train_scores
    ]

    res.loc[:, "train_score"] = train_score.transpose().to_numpy()

    return res


def score_box_plot(score, metric, aspect=1, rot=0):
    g = sns.catplot(aspect=aspect)

    g = sns.scatterplot(
        data=score,
        x="classifier",
        y="train_" + metric,
        markers="X",
        label=metric + " на тренировочном наборе",
    )

    g = sns.boxplot(
        data=score, x="classifier", y=metric, color="white", showmeans=True
    )

    g.tick_params(axis="x", rotation=rot)
    g.set_xlabel("Классификатор")
    g.set_ylabel(metric)
    g.set_title(
        f"Значения метрики {metric} по результатм"
        "\nкросс-валидации для разных классификаторов"
    )
    sns.move_legend(g, "lower right")

    plt.show()


def evaluate_model(
    estimator, X, y, label, metrics, ax, cv, n_jobs, method="predict_proba"
):
    """Печатает диаграмму размаха для результатов кросс-валидации указанной
    метрики. Отмечает на диаграмме результат, полученный на тренировочном
    наборе данных.
    """

    cols = ["classifier"] + metrics + ["train_" + value for value in metrics]

    scores = pd.DataFrame(columns=cols)

    for metric in metrics:
        scores[metric] = get_cv_scores(
            estimator, X, y, score=metric, cv=cv, n_jobs=n_jobs
        )
        scores["train_" + metric] = get_train_score(
            estimator, X, y, score=metric
        )

    y_scores = get_cv_pred(estimator, X, y, method=method, cv=cv, n_jobs=n_jobs)
    ax = plot_roc_curve(y, y_scores, label=label, ax=ax)

    scores["classifier"] = label

    return scores, ax


def compare_models(
    classifiers,
    methods,
    labels,
    X,
    y,
    cv,
    n_jobs,
    box_plot_aspect=1.0,
    box_plot_xrot=0,
):
    all_scores = pd.DataFrame()

    fig, ax = plot_roc_curve_for_random_clf()

    for i, (classifier, method, label) in enumerate(
        zip(classifiers, methods, labels)
    ):
        print(f"{i+1}. {label}")

        metrics_list = ["accuracy", "f1", "roc_auc"]

        scores, ax = evaluate_model(
            estimator=classifier,
            X=X,
            y=y,
            metrics=metrics_list,
            label=label,
            method=method,
            ax=ax,
            cv=cv,
            n_jobs=n_jobs,
        )

        print(f"Значения метрик для {label}")
        display(                                                    # noqa: F821
            agg_scores(scores, label=label, score_names=metrics_list)
        )

        all_scores = pd.concat([all_scores, scores])

    plt.show()

    for metric in metrics_list:
        score_box_plot(
            all_scores, metric=metric, aspect=box_plot_aspect, rot=box_plot_xrot
        )
        plt.show()
